### PR TITLE
[Onyx-742] Make PhysicalPlan scalable

### LINKED ISSUE #742: Make PhysicalPlan scalable
Currently, the size of PhysicalPlan is proportional to not only the *complexity* of the job, but also the *volume* of the input data, by repeatedly generating TaskGroups and TaskGroup-internal DAGs. By deduplicating (almost) same internal DAGs, we may greatly reduce memory pressure on Onyx master.

### PR DESCRIPTION
**Pull request by sanha**
This PR:
- Removes physical-level information from `Task`.
- Moves physical-level information from `TaskGroup` (index, for example) to `ScheduledTaskGroup`.

Resolves #742.

### CODE REVIEW

File: runtime/common/src/main/java/edu/snu/onyx/runtime/common/RuntimeIdGenerator.java [hunk 1]
```diff
@@ -23,13 +23,13 @@
  */
 public final class RuntimeIdGenerator {
   private static AtomicInteger physicalPlanIdGenerator = new AtomicInteger(0);
-  private static AtomicInteger taskIdGenerator = new AtomicInteger(0);
-  private static AtomicInteger taskGroupIdGenerator = new AtomicInteger(0);
   private static AtomicInteger executorIdGenerator = new AtomicInteger(0);
   private static AtomicLong messageIdGenerator = new AtomicLong(1L);
   private static AtomicLong resourceSpecIdGenerator = new AtomicLong(0);
   private static String blockPrefix = "Block-";
```
Comment (johnyangk) [inline runtime/common/src/main/java/edu/snu/onyx/runtime/common/RuntimeIdGenerator.java]: L29-L32
Make the variables `final`

File: runtime/common/src/main/java/edu/snu/onyx/runtime/common/plan/physical/PhysicalPlanGenerator.java [hunk 2]
```diff
@@ -181,90 +186,82 @@ private PhysicalPlanGenerator(@Parameter(JobConf.DAGDirectory.class) final Strin
...
       });
-      final PhysicalStage physicalStage = physicalStageBuilder.build();
-      physicalDAGBuilder.addVertex(physicalStage);
 
+      // Create the task group to add for this stage.
+      final TaskGroup newTaskGroup = new TaskGroup(stage.getId(), stageInternalDAGBuilder.build(), containerType);
+      final PhysicalStage physicalStage =
+          new PhysicalStage(stage.getId(), newTaskGroup, stageParallelism, stage.getScheduleGroupIndex());
+
+      irVertexTaskMap.clear();
```
Comment (johnyangk) [inline runtime/common/src/main/java/edu/snu/onyx/runtime/common/plan/physical/PhysicalPlanGenerator.java]: Remove this and move `irVertexTaskMap` into the loop?

File: runtime/common/src/main/java/edu/snu/onyx/runtime/common/plan/physical/TaskGroup.java [hunk 3]
```diff
@@ -27,53 +27,31 @@
  * A stage contains a list of TaskGroups whose length corresponds to stage/operator parallelism.
  */
 public final class TaskGroup implements Serializable {
```
Comment (johnyangk) [inline runtime/common/src/main/java/edu/snu/onyx/runtime/common/plan/physical/TaskGroup.java]: Probably unrelated to this PR, but it'd be nice if this class can be removed.

Its variables can be removed/moved like the following
* stageId -> remove, and instead provide PhysicalStage#getId
* taskDAG -> move into PhysicalStage
* containerType -> probably better to be tracked separately in a scheduler component

Can you file an issue, if this makes sense?

File: runtime/executor/src/main/java/edu/snu/onyx/runtime/executor/datatransfer/DataTransferFactory.java [hunk 4]
```diff
@@ -25,49 +25,59 @@ public DataTransferFactory(@Parameter(JobConf.HashRangeMultiplier.class) final i
...
+   * @param srcTask     the {@link Task} that outputs the data to be written.
+   * @param srcTaskIdx  the index of the source task.
+   * @param runtimeEdge that connects the srcTask to the tasks belonging to dstIRVertex.
+   * @return the {@link OutputWriter} created.
+   */
   public OutputWriter createLocalWriter(final Task srcTask,
+                                        final int srcTaskIdx,
                                         final RuntimeEdge<?> runtimeEdge) {
-    return createWriter(srcTask, null, runtimeEdge);
+    return createWriter(srcTask, srcTaskIdx, null, runtimeEdge);
```
Comment (johnyangk) [inline runtime/executor/src/main/java/edu/snu/onyx/runtime/executor/datatransfer/DataTransferFactory.java]: `null` doesn't look good. Can you add a TODO?

File: runtime/executor/src/main/java/edu/snu/onyx/runtime/executor/datatransfer/DataTransferFactory.java [hunk 5]
```diff
@@ -25,49 +25,59 @@ public DataTransferFactory(@Parameter(JobConf.HashRangeMultiplier.class) final i
...
-   * @param runtimeEdge      that connects the tasks belonging to srcRuntimeVertex to dstTask.
+   * @param dstTaskIdx  the index of the destination task.
+   * @param runtimeEdge that connects the tasks belonging to srcRuntimeVertex to dstTask.
    * @return the {@link InputReader} created.
    */
-  public InputReader createLocalReader(final Task dstTask,
+  public InputReader createLocalReader(final int dstTaskIdx,
                                        final RuntimeEdge runtimeEdge) {
-    return createReader(dstTask, null, runtimeEdge);
+    return createReader(dstTaskIdx, null, runtimeEdge);
```
Comment (johnyangk) [inline runtime/executor/src/main/java/edu/snu/onyx/runtime/executor/datatransfer/DataTransferFactory.java]: Ditto.

File: runtime/executor/src/main/java/edu/snu/onyx/runtime/executor/datatransfer/OutputWriter.java [hunk 6]
```diff
@@ -45,14 +45,14 @@
   public OutputWriter(final int hashRangeMultiplier,
                       final int srcTaskIdx,
                       final String srcRuntimeVertexId,
-                      @Nullable final IRVertex dstRuntimeVertex, // Null if it is not a runtime vertex.
+                      @Nullable final IRVertex dstIrVertex, // Null if it is not a IR vertex.
```
Comment (johnyangk) [inline runtime/executor/src/main/java/edu/snu/onyx/runtime/executor/datatransfer/OutputWriter.java]: Ditto.

File: tests/src/test/java/edu/snu/onyx/tests/runtime/master/scheduler/RoundRobinSchedulingPolicyTest.java [hunk 7]
```diff
@@ -101,118 +102,100 @@ public void checkScheduleTimeout() {
...
-    return new ScheduledTaskGroup("TestPlan", taskGroup, Collections.emptyList(), Collections.emptyList(),
-        MAGIC_SCHEDULE_ATTEMPT_INDEX);
+  /**
+   * Wrap a {@link TaskGroup} into {@link ScheduledTaskGroup}s.
+   *
+   * @param parallelism how many scheduled task group will be generated.
+   * @param taskGroup   the task group to schedule.
+   * @return the wrapped scheduled task groups.
+   */
+  private List<ScheduledTaskGroup> wrap(final int parallelism,
```
Comment (johnyangk) [inline tests/src/test/java/edu/snu/onyx/tests/runtime/master/scheduler/RoundRobinSchedulingPolicyTest.java]: converToScheduledTaskGroups

File: tests/src/test/java/edu/snu/onyx/tests/runtime/master/scheduler/SingleTaskGroupQueueTest.java [hunk 8]
```diff
@@ -415,4 +404,15 @@ public void testContainerTypeAwareness() throws Exception {
 
     countDownLatch.await();
   }
+
+  /**
+   * Schedule the task groups in a physical stage.
+   * @param stage the stage to schedule.
+   */
+  private void scheduleTaskGroupsInStage(final PhysicalStage stage) {
```
Comment (johnyangk) [inline tests/src/test/java/edu/snu/onyx/tests/runtime/master/scheduler/SingleTaskGroupQueueTest.java]: scheduleStage

File: tests/src/test/java/edu/snu/onyx/tests/runtime/master/scheduler/SingleTaskGroupQueueTest.java [hunk 9]
```diff
@@ -377,14 +369,11 @@ public void testContainerTypeAwareness() throws Exception {
 
     final CountDownLatch countDownLatch = new CountDownLatch(2);
 
-    // First schedule the children TaskGroups (since it is push).
+    // First schedule the child TaskGroups (since it is push).
```
Comment (johnyangk) [inline tests/src/test/java/edu/snu/onyx/tests/runtime/master/scheduler/SingleTaskGroupQueueTest.java]: child TaskGroups -> child TaskGroup (drop the s)

File: runtime/common/src/main/java/edu/snu/onyx/runtime/common/plan/physical/TaskGroup.java (hunk 3 above)
Comment (sanha) [reply runtime/common/src/main/java/edu/snu/onyx/runtime/common/plan/physical/TaskGroup.java]: I like it. I'll file the issue.

**Comment by sanha** [comment]
!rebuild

**Pull request merged**